### ISSUE TITLE
Could not properly parse responses to match requests in batch

### ISSUE THREAD

**Issue by JackHu88** [SharePoint/PnP-JS-Core#365]
#### Category
[ ] Enhancement

[ ] Bug

[x ] Question

#### Expected / Desired Behavior / Question
I'm refering pnp.js on my SPO project, i want add many items one time to my list using the batch function. there are 6 items in my json data, but only 5 items was added, then throw an error below:

#### Observed Behavior
_Error: Could not properly parse responses to match requests in batch.
    at new BatchParseException (pnp.js:3256)
    at pnp.js:2661_

#### Steps to Reproduce
Code below:
```
var promises=[];
var batch = $pnp.sp.createBatch();
var list=$pnp.sp.web.lists.getByTitle("Targets");

$.each(jsonData,function(index,item){					
	var fields={};				
	fields.CaseID=jsonData.caseid;
	fields.Market_x0020_P_x002d_file=jsonData.mpfile;						
	fields.Target_x0020_Name_x0028_CN_x0029=jsonData.nameCN;
	fields.Target_x0020_Name_x0028_EN_x0029=jsonData.nameEN;
	fields.Title=jsonData.tpfile;
	fields.Open_x0020_Date=jsonData.openDate;
	fields.Province=jsonData.province;
	fields.City=jsonData.city;
	fields.Group=jsonData.group;
	fields.Division=jsonData.division;
	fields.Product=jsonData.product;
	fields.Description=jsonData.description;
	fields.Agent=jsonData.agent;
	fields.Value_x0020_of_x0020_Seized_x002=jsonData.value;
	fields.Action=jsonData.action;
	fields.Seized_x0020_Products=jsonData.seizedProducts;
					
	promises.push(list.items.add(fields));	
})						
Promise.all(promises).then(function () {
	console.log("prepare completed");
})		
batch.execute().then(function() {
	console.log("completed");
}).catch(function(err){
	console.log(err);
})
```

**Comment by patrick-rodgers** [comment]
I don't see the item adding operations in your example but I suspect this is being caused by a race condition which I believe to be fixed now in the dev branch. To get around it you can use the pattern [here](https://github.com/SharePoint/PnP-JS-Core/wiki/Working-With:-Items#add-multiple-items) that gets the list item type name and then does the add operations.

**Comment by JackHu88** [comment]
Thanks @patrick-rodgers , 

Above code **promises.push(list.items.add(fields));**  is the add option.

Does your example [here ](https://github.com/SharePoint/PnP-JS-Core/wiki/Working-With:-Items#add-multiple-items) can add 100+ items in a batch?

**Comment by patrick-rodgers** [comment]
Ah, I see what you are doing now. As for 100+ items I am not sure. There are no published limits as to what the batching can handle, but you should test it out. I did just a quick test and the below worked for me. This is with the fix in the dev branch for that race condition I mentioned above.

```TypeScript
    let batch = pnp.sp.web.createBatch();
    let list = pnp.sp.web.lists.getByTitle("Test345-2");

    for (let i = 0; i < 100; i++) {

        list.items.inBatch(batch).add({
            Title: Util.getGUID(),
        });
    }

    batch.execute().then(_ => {

        console.log("Batch is done.");

    });
```

**Comment by JackHu88** [comment]
Hi @patrick-rodgers , I didn't find the pnp.js in the dev branch, i referring the pnp.js you published 6 days ago on my SPO environment. I still have the error below, only 99 items was created. 
I tested 5 times but only once created successfully no error.

> Uncaught (in promise) Error: Could not properly parse responses to match requests in batch.

```
var batch = $pnp.sp.web.createBatch();
var list = $pnp.sp.web.lists.getByTitle("BatchTest");
for (var i = 0; i < 100; i++) {
    list.items.inBatch(batch).add({
        Title: "Test"+i
    })
}
batch.execute().then(_ => {
    console.log("Batch is done.");
});
```
I see someone mentioned the batch request will be processed by order, but it seems some items didn't  created by order. some items order like below：
Test81
Test82
Test84
Test83
Test74
Test85

**Comment by JackHu88** [comment]
I have tested the list.getListItemEntityTypeFullName() can process 100+ items, but i have an new question now. I want add many users to a site, so i running the code below to test it, but when i check the request header, the ensureuser api was requested 3 times not processed in a batch.

```
var batch = $pnp.sp.web.createBatch();

$pnp.sp.web.inBatch(batch).ensureUser("jhu1@xxx.com").then(function (user){
    var userId= user.Id;	
    console.log(userId);
})
$pnp.sp.web.inBatch(batch).ensureUser("jhu2@xxx.com").then(function (user){
    var userId= user.Id;	
    console.log(userId);
})
$pnp.sp.web.inBatch(batch).ensureUser("jhu3@xxx.com").then(function (user){
    var userId= user.Id;	
    console.log(userId);
})

batch.execute().then(d => console.log("Done"));
```
![user](https://cloud.githubusercontent.com/assets/26239484/24031818/a65917ee-0b1f-11e7-925b-c36f02f621ac.png)

**Comment by JackHu88** [comment]
Hi @patrick-rodgers , could you please help me have a look this issue?

**Comment by patrick-rodgers** [comment]
Hi @JackHu88, always trying to help but understand not really able to provide on-demand support.

For your first question about only 99 items being created, I believe I mentioned above to [have a look at this pattern](https://github.com/SharePoint/PnP-JS-Core/wiki/Working-With:-Items#add-multiple-items) and get the list item type name before you are adding the items to the batch and then pass that to the add method.

For your second could you please open another issue for that so we can have a look? Need to keep the threads to a single topic for easier tracking and helps to ensure things aren't missed. But there might be a bug in the ensureUser method so will want to have a look.

Thanks!

**Comment by JackHu88** [comment]
Closing now, i opened an new issue [#380](https://github.com/SharePoint/PnP-JS-Core/issues/380) for my second question.
